### ISSUE TITLE
Panic: error parsing from slice: could not convert slice to array

### ISSUE THREAD

**Issue by wcampbell0x2a** [sharksforarms/deku#224]
Adding the following example in deku causes this error:
```
---- it_works stdout ----
thread 'it_works' panicked at 'called `Result::unwrap()` on an `Err` value: Parse("error parsing from slice: could not convert slice to array")', tests/test_struct.rs:228:50
```

```diff
diff --git a/tests/test_struct.rs b/tests/test_struct.rs
index 062d006..66c015a 100644
--- a/tests/test_struct.rs
+++ b/tests/test_struct.rs
@@ -167,3 +167,64 @@ fn test_raw_identifiers_struct() {
     let ret_write: Vec<u8> = ret_read.try_into().unwrap();
     assert_eq!(test_data, ret_write);
 }
+
+#[test]
+fn it_works() {
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    pub struct Packet {
+        pub data: Data,
+    }
+
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    pub struct Data {
+        pub one: One,
+        pub two: Two,
+    }
+
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    #[deku(type = "u8", bits = "2")]
+    pub enum One {
+        Start = 0,
+        Go = 1,
+        Stop = 2,
+    }
+
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    #[deku(type = "u8", bits = "4")]
+    pub enum Two {
+        #[deku(id = "0b0000")]
+        Put(Op1),
+        #[deku(id = "0b0001")]
+        Store(Op1),
+        #[deku(id = "0b0010")]
+        Load(Op2),
+        #[deku(id = "0b0011")]
+        Allocate(Op2),
+    }
+
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    pub struct Op1 {
+        #[deku(bits = "2")]
+        pub i: u8,
+        #[deku(bits = "4")]
+        pub o: u8,
+    }
+
+    #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+    pub struct Op2 {
+        #[deku(bits = "8")]
+        pub w: u8,
+        #[deku(bits = "6")]
+        pub j: u8,
+    }
+
+    let mut packet = Packet {
+        data: Data {
+            one: One::Start,
+            two: Two::Load(Op2 { w: 1, j: 2 }),
+        },
+    };
+    let bytes = packet.to_bytes().unwrap();
+    let packet = Packet::from_bytes((&bytes, 0)).unwrap();
+}
```

**Comment by constfold** [comment]
It seems like the generated code call `try_info` to convert slice to a fixed-size array with false assumption. But I don't find any such uses in deku macro, so maybe it happened in here:
https://github.com/sharksforarms/deku/blob/9cc9ba1f8facb58745465efc2eaa0ed655eea3bb/src/impls/primitive.rs#L84-L88
Could you repelace it with `try_info().unwrap()` and try this test again? If the guess is true, a better message will appear.

BTW, this issue also shows another problem: It is awful to debug when the generated code panic or not work as expected because the error context is totally missed and all codes are hidden behind the macro.

**Comment by wcampbell0x2a** [comment]
:worried: 
```
thread 'it_works' panicked at 'called `Result::unwrap()` on an `Err` value: TryFromSliceError(())', src/impls/primitive.rs:208:1
```

**Comment by constfold** [comment]
According to panic message, it was due to the implementation of `u8`.
Unfortunately I doesn't have rust env installed in my current PC, so I can't debug it. I will try to work on this a few days after if it hasn't been resolved.

**Comment by sharksforarms** [comment]
I'll have time to take a look tomorrow

**Comment by sharksforarms** [comment]
I believe I made some false assumptions here in terms of alignment.

```diff
diff --git a/src/impls/primitive.rs b/src/impls/primitive.rs
index fb58dad..0f82a61 100644
--- a/src/impls/primitive.rs
+++ b/src/impls/primitive.rs
@@ -35,10 +35,12 @@ macro_rules! ImplDekuTraits {
                 let value = if pad == 0 && bit_slice.len() == max_type_bits {
                     // if everything is aligned, just read the value

+                   // Bits:  [00, 000001]
                     let bytes: &[u8] = bit_slice.as_raw_slice();
+                   // Bytes: [8, 4] <-- two bytes

                     // Read value
                     if input_is_le {
+                       // ERR: 2 bytes doesn't fit in u8
                         <$typ>::from_le_bytes(bytes.try_into()?)
                     } else {
                         <$typ>::from_be_bytes(bytes.try_into()?)

```

This line here specifically was added to prevent re-creating a BitVec, this is not correct. We need to make sure somehow that the bitslice is contiguous, @myrrlyn any thoughts?

Wondering if this is a good fix:

```diff
diff --git a/src/impls/primitive.rs b/src/impls/primitive.rs
index fb58dad..e205959 100644
--- a/src/impls/primitive.rs
+++ b/src/impls/primitive.rs
@@ -28,11 +28,14 @@ macro_rules! ImplDekuTraits {
                     return Err(DekuError::Incomplete(crate::error::NeedSize::new(bit_size)));
                 }

                 let (bit_slice, rest) = input.split_at(bit_size);

                 let pad = 8 * ((bit_slice.len() + 7) / 8) - bit_slice.len();

-                let value = if pad == 0 && bit_slice.len() == max_type_bits {
+                let value = if pad == 0 && bit_slice.as_raw_slice().len()*8 == max_type_bits  {
                     // if everything is aligned, just read the value

                     let bytes: &[u8] = bit_slice.as_raw_slice();
```

https://github.com/sharksforarms/deku/blob/5fc56a232fa564ed828c7079c9c1cb2d90f8cd5b/src/impls/primitive.rs#L35

**Comment by sharksforarms** [comment]
pushed a fix in 0.12.2
